core-rust/state-manager/src/protocol/protocol_updates/protocol_update_progress.rs

```diff
@@ -19,31 +20,40 @@ impl ProtocolUpdateProgress {
     /// Inspects the database in order to resolve the current Protocol Update progress.
     pub fn resolve(database: &impl QueryableProofStore) -> Self {
         let Some(latest_proof) = database.get_latest_proof() else {
-            return ProtocolUpdateProgress::NotUpdating;
+            return ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
+                protocol_version_name: ProtocolVersionName::babylon(),
+            };
         };
 
-        match &latest_proof.origin {
-            LedgerProofOrigin::Genesis { .. } => ProtocolUpdateProgress::NotUpdating,
-            LedgerProofOrigin::Consensus { .. } => {
-                if let Some(latest_proof_protocol_version) =
-                    latest_proof.ledger_header.next_protocol_version
-                {
-                    ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
-                        protocol_version_name: ProtocolVersionName::of_unchecked(
-                            latest_proof_protocol_version,
-                        ),
-                    }
-                } else {
-                    ProtocolUpdateProgress::NotUpdating
-                }
-            }
+        if let Some(latest_proof_protocol_version) =
+            &latest_proof.ledger_header.next_protocol_version
+        {
+            return ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
+                protocol_version_name: latest_proof_protocol_version.clone(),
+            };
+        }
+
+        match latest_proof.origin {
+            LedgerProofOrigin::Consensus { .. } => ProtocolUpdateProgress::NotUpdating,
             LedgerProofOrigin::ProtocolUpdate {
                 protocol_version_name,
+                config_hash: _,
+                batch_name: _,
+                batch_group_index,
+                batch_group_name: _,
                 batch_index,
-            } => ProtocolUpdateProgress::UpdateInProgress {
-                protocol_version_name: protocol_version_name.clone(),
-                last_batch_index: *batch_index as usize,
-            },
+                is_end_of_update,
+            } => {
+                if is_end_of_update {
+                    ProtocolUpdateProgress::NotUpdating
+                } else {
```

Comment:
More of a question: during our sync, you mentioned that there would be trigger after `end_of_update`. If a node stops right after committing the last element of the last batch, how would the triggering be recovered as we set progress to `NotUpdating`?

Reply:
That's correct.

And the answer is that such proofs will have both `ledger_header.next_protocol_version` and `is_end_of_update` set, and about 10 lines above, we have this code:
```rust
        if let Some(latest_proof_protocol_version) =
            &latest_proof.ledger_header.next_protocol_version
        {
            return ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
                protocol_version_name: latest_proof_protocol_version.clone(),
            };
        }
```

The `next_protocol_version` is set in `SystemCommitRequestFactory::create_proof` as a result of:
```
        let end_state = series_executor.finalize_series(batch_details.to_batch_situation());
        SystemPrepareResult::from_committed_series(committed_transactions, end_state)
```
And the `series_executor.finalize_series` ultimately gets it from:
```
        let next_protocol_version = self
            .protocol_state
            .check_for_update_trigger_at_end_of_batch(
                self.epoch_change.as_ref(),
                self.state_version,
                batch_situation,
            );
```